src/chain/auth.sw

```diff
@@ -1,28 +1,39 @@
 library auth;
+//! Functionality for determining who is calling an ABI method
 
-// this can be a generic option when options land
-enum Caller {
-    Some: b256,
-    None: (),
+use ::result::Result;
+use ::address::Address;
+use ::contract_id::ContractId;
+
+pub enum AuthError {
+    ContextError: (),
+}
+
+pub enum Sender {
+    Address: Address,
+    Id: ContractId,
 }
 
-/// Returns `true` if the caller is external.
+/// Returns `true` if the caller is external (ie: a script or predicate).
 pub fn caller_is_external() -> bool {
     asm(r1) {
         gm r1 i1;
         r1: bool
     }
 }
 
-pub fn caller() -> Caller {
-    // if parent is not external
+/// Returns a Result::Ok(Address) or Result::Error.
+// NOTE: Currently only returns the Ok variant of result if the parent context is Internal.
+pub fn msg_sender() -> Result<Sender, AuthError> {
     if !caller_is_external() {
```

Comment:
I would find this more readable if I dropped the `!` operator and swap the cases, so:
```
if caller_is_external {} else {}
```
The `!`, while minor, adds needless complication.

Reply:
i agree, non-negated logic, or as my college professor would say "backwards-ass ifs," add mental overhead.